Docker development QoL improvements

jrmehle opened the pull request:
**NOTE: DO NOT discuss internal CommitChange information in your PR; this PR will be public.
Link back to the issue in the Tix repo when you need to do that.**

Made a number of quality of life improvements to local development with Docker containers. 
* Removed deprecated version line in docker-compose.yml
* Added tty and std_open config to support pry debugging
* Allow setting of PORT and BASE_RELEASE env vars in docker-compose.yml
* Use tmpfs for pids folder
* Use consistent syntax for ENV
* Install only requested packages and not any recommended by package manager
* Config for logging to STDOUT through `RAILS_LOG_TO_STDOUT` ENV var
* Preserve database data between container restarts
* jemalloc and YJIT support for local dev
* consistently use port 5000 by default for Rails server/web container but allow overriding in .env
* move jobs worker to its own container
* Created cc-run command to make interacting with containers easier
* Converted single instance of a tab to spaces

jrmehle committed:
Remove deprecated version line from docker-compose

jrmehle committed:
Allow for pry debugging

jrmehle committed:
Allow for setting of port via env var

jrmehle committed:
Allow setting of base release via env var

jrmehle committed:
Use tmpfs for pids folder

Should remove need to clean up pid files

jrmehle committed:
tab => spaces

jrmehle committed:
Consistent ENV syntax

Both with `=` and without are valid, but the latter can cause issues with multi-line values so let's be explicit and use `=` .

jrmehle committed:
no bin/dev start

It seems from local testing that we want `bin/dev` and not `bin/dev start` here.

This may have been from a previous iteration where `foreman start` was called directly.

jrmehle committed:
Log Rails to STDOUT using RAILS_LOG_TO_STDOUT

jrmehle committed:
Don't install recommended packages

Keep containers as small as possible by only including requested packages and nothing the package manager suggests.

Also reformatted long line of packages names into list with one per line for readability.

jrmehle committed:
Persist database files while db container is down

jrmehle committed:
Enable YJIT and jemalloc in Docker

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Lessen rspec output noise

As rspec tests ran, logs were spewing into STDOUT where rspec was also displaying its progress making it impossible to tell where it was.

jrmehle committed:
Clean up spacing and unused code

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Quiet rack-timeout logs in dev

wwahammy reviewed (commented):
Thanks @jrmehle, it looks like things are moving in the correct direction.

As an aside, it would probably make sense to have a Github workflow to test this docker compose setup. It'd be nice to know if it builds and runs once we get to that point.

jrmehle committed:
Switch base release to bookworm

The install seems to fail on mismatched postgresql-client-16 fails to install in bullseye

jrmehle committed:
Run rails server on port 5000 but allow .env override

Since .env is not checked in. Each developer can set PORT in .env to control which port the Rails server starts up on.

jrmehle committed:
Move lograge to only be used in deployed envs

The structured logging is useful in staging and production (currently Heroku), but less so in development. Further, Rails logging has improved drastically since lograge was first introduced to the project.

jrmehle committed:
Don't make timeout logging depend on IDE_PROCESS_DISPATCHER

jrmehle committed:
Pull out unneeded change to puma port config

At one point had a bind call in the puma config with the port on it. Created `tcp_port` so the env var didn't need to be read twice. But found out the bind that was being specified was essentially the same as default and removed it. This should have been cleaned up at the time but was missed.

wwahammy commented on docker-compose.yml:
Not immediate but I have a strong preference that we remove the forwarded ports for the db. It just causes conflicts when you're starting up (what if two things are running!). You can properly access the database by using `bin/rails dbconsole -p`

jrmehle committed:
Revert back to setting port in .foreman

Ran into a buildtime issue where foreman didn't have a PORT which resulted in puma starting on a random port. Because of this the port mapping in Docker was not correct and the web container became unreachable via browser.

jrmehle committed:
Break out separate worker container

This also has the benefit of separating logs so the periodic activity of DelayedJob doesn't mess with the view of the Rails logs.

jrmehle committed:
Revert foreman port

jrmehle committed:
Provide to run worker as normal outside Docker

When using Docker, it is useful to have the job logs separate and so it makes sense to run the job worker in a separate container. For a dev running the project on bare metal (without Docker), they should have an easy one-command process as it had been with bin/dev prior. This maintains that by making bin/dev switch which Procfile it uses based on an ENV var.

jrmehle committed:
Add a run file to improve Docker dev experience

Big wheel keep on turnin'
Proud Mary keep on burnin'
Rollin', rollin', rollin' on the river

jrmehle committed:
More ccr commands

* Added compatibility for different versions docker-compose.
* Added passthrough standardrb command.
* Added setup command to run Rails setup.
* Added restart and rebuild commands to manage docker containers.

jrmehle committed:
Add db-restore command to ccr

Restores from a local copy of the DB

jrmehle committed:
Install imagemagick for handling images

This project makes use of minimagick to resize images and needs the `convert` command from imagemagick to do that.

jrmehle commented on config/puma.rb:
```diff
@@ -62,12 +62,16 @@
-  ActiveRecord::Base.establish_connection
+if env != "development"
+  on_worker_boot do
+    ActiveRecord::Base.establish_connection
+  end
 end
 
-before_fork do
-  Barnes.start
+if env != "development"
+  before_fork do
+    Barnes.start
+  end
```
Puma doesn't like/need to have multiple workers in development. Setting `WEB_CONCURRENCY` to 0 in development puts it into single-worker mode. But then puma starts complaining about `on_worker_boot` and `before_fork` in that mode.

jrmehle commented:
> Thanks @jrmehle, it looks like things are moving in the correct direction.
> 
> As an aside, it would probably make sense to have a Github workflow to test this docker compose setup. It'd be nice to know if it builds and runs once we get to that point.

Is there an example you had in mind? I searched around for a bit and found https://github.com/adambirds/docker-compose-action. Do you see workflow action check as a blocker?

wwahammy commented:
> Is there an example you had in mind? I searched around for a bit and found https://github.com/adambirds/docker-compose-action. Do you see workflow action check as a blocker?

If we intend folks to use Docker compose as the default dev setup, I don't know if I'd consider it a blocker but it's close. It's too darn easy to make mistakes that pass checks otherwise but actually break Docker setup. I don't think the full test suite needs to run, just the main development setup (minus downloading the live data). If that passes and nothing breaks, I'd say that's enough for CI.

Also, I think that Github action seems reasonable.

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Add check of docker compose build for development

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Merge branch 'supporter_level_goal' into jaredmehle/fix/docker-development

jrmehle committed:
Increase shared memory for database container

This fixes errors that sometimes occur on big queries in the Rails backend.

jrmehle committed:
Merge branch 'jaredmehle/fix/docker-development' of github.com:CommitChange/houdini into jaredmehle/fix/docker-development

jrmehle committed:
Fix syntax

jrmehle committed:
Change where gems are stored in Docker

jrmehle committed:
Update docker build check checkout version

wwahammy committed:
Stop forwarding ports unnecessarily

wwahammy committed:
Correct the usage of docker compose

wwahammy reviewed (approved):
Looks good to me @jrmehle!
